pkg/security/admission/admission.go

```diff
@@ -0,0 +1,317 @@
+	"io"
+
+	kadmission "github.com/GoogleCloudPlatform/kubernetes/pkg/admission"
+	kapi "github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api/errors"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client/cache"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/fields"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/runtime"
+	scc "github.com/GoogleCloudPlatform/kubernetes/pkg/securitycontextconstraints"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/serviceaccount"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/watch"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/auth/user"
+	allocator "github.com/openshift/origin/pkg/security"
+	"github.com/openshift/origin/pkg/security/uid"
+)
+
+func init() {
+	kadmission.RegisterPlugin("SecurityContextConstraint", func(client client.Interface, config io.Reader) (kadmission.Interface, error) {
+		return NewConstraint(client), nil
+	})
+}
+
+type constraint struct {
+	*kadmission.Handler
+	client client.Interface
+	store  cache.Store
+}
+
+var _ kadmission.Interface = &constraint{}
+
+func NewConstraint(kclient client.Interface) kadmission.Interface {
+	store := cache.NewStore(cache.MetaNamespaceKeyFunc)
+	reflector := cache.NewReflector(
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return kclient.SecurityContextConstraints().List(labels.Everything(), fields.Everything())
+			},
+			WatchFunc: func(resourceVersion string) (watch.Interface, error) {
+				return kclient.SecurityContextConstraints().Watch(labels.Everything(), fields.Everything(), resourceVersion)
+			},
+		},
+		&kapi.SecurityContextConstraints{},
+		store,
+		0,
+	)
+	reflector.Run()
+
+	return &constraint{
+		Handler: kadmission.NewHandler(kadmission.Create, kadmission.Update),
+		client:  kclient,
+		store:   store,
+	}
+}
+
+// Admit determines if the pod should be admitted based on the requested security context
+// and the available SCCs.
+//
+// 1.  Find all the SCCs that the service account or user has access to
+// 2.  Full resolve each SCC
+//     1.  Determine if the user/selinux strategies requires pre-allocated values
+//     2.  Retrieve pre-allocated values from the namespace and set them on the strategies
+// 3.  Create the set of providers from the fully configured SCCs
+// 4.  Generate and validate each container's SC against the set of SCC providers.  If any container
+//     fails to validate then the entire request is rejected
+func (c *constraint) Admit(a kadmission.Attributes) error {
+	if a.GetResource() != string(kapi.ResourcePods) {
+		return nil
+	}
+
+	pod, ok := a.GetObject().(*kapi.Pod)
+	// if we can't convert then we don't handle this object so just return
+	if !ok {
+		return nil
+	}
+
+	// all pods are required to have a service account
+	serviceAccount, err := getServiceAccount(c.client, a.GetNamespace(), pod)
```

Comment:
if there's no service account, shouldn't it just use the constraints for the calling user? could even check if len(pod.spec.serviceaccount) > 0 first. keep mirror pods in mind as well... those won't have service accounts and mutating their specs will make the kubelet angry